### PR TITLE
Try using ensurepip instead of get-pip

### PR DESCRIPTION
**Pull request by joerick**
Eliminating get-pip.py removes a vendored dependency.

### CODE REVIEW

File: cibuildwheel/windows.py [hunk 1]
```diff
@@ -174,9 +176,24 @@ def setup_python(
...
+    # make sure pip is installed and available on PATH
     if not (installation_path / "Scripts" / "pip.exe").exists():
-        call(["python", get_pip_script, *dependency_constraint_flags], env=env, cwd="C:\\cibw")
+        # perhaps pip is installed, but not available as 'pip.exe'...
+        pip_is_installed = (
+            call(
+                ["python", "-m", "pip", "--version"], env=env, cwd="C:\\cibw", check=False
+            ).returncode
+            == 0
+        )
```
Comment (joerick) [inline cibuildwheel/windows.py:189]: This bit is kinda clumsy, as is adding the return type on `call()`... maybe the following is better?

```suggestion
        try:
            call(["python", "-m", "pip", "--version"], env=env, cwd="C:\\cibw")
        except subprocess.CalledProcessError:
            pip_is_installed = False
        else:
            pip_is_installed = True
```

Comment (joerick) [reply cibuildwheel/windows.py:189]: Yeah I'm gonna change it…

**Comment by joerick** [comment]
Hmm. It was working, but the 3.8.10 update broke it on mac! I wonder what changed...

**Comment by YannickJadoul** [comment]
Cross-linking to #210. Great that it does now seem to be working! :-)

**Comment by joerick** [comment]
This is ready to go I think. To explain what's going on here, we need to ensure that pip is available on PATH as `pip`, because user options like BEFORE_BUILD will likely use that. But distributions differ in that respect, some have `pip` available, some only as `pip3`, some only have pip in site-packages for use as `python -m pip`. 

So, to ensure that `pip` is available, you run `python -m ensurepip --default-pip`. But! If pip is installed, ensurepip does nothing, so it doesn't act on `--default-pip`. So the only way to get `--default-pip` to work is to make sure that ensurepip is actually _installing_ pip! So that's the why behind the little dance we're doing here. 

This seems okay, but alternative ideas welcome!

File: cibuildwheel/macos.py [hunk 2]
```diff
@@ -232,8 +231,29 @@ def setup_python(
...
-    call(["python", get_pip_script, *dependency_constraint_flags], env=env, cwd="/tmp")
+    # ensure pip is installed
+    if not (installation_bin_path / "pip").exists():
+        # perhaps pip is installed, but not available as 'pip'...
+        try:
+            call(["python", "-m", "pip", "--version"], env=env, cwd="/tmp")
+        except subprocess.CalledProcessError:
+            pip_is_installed = False
+        else:
+            pip_is_installed = True
```
[the remark this answers is not in the document]
Comment (joerick) [reply cibuildwheel/macos.py:242]: haha, yes that was my first instinct, too. You can see it here:
https://github.com/joerick/cibuildwheel/pull/665/commits/ef9163550e375127f25e6fa04a58a01b58e23f8e 

However, I ended up (https://github.com/joerick/cibuildwheel/pull/665#discussion_r628860733) changing that to the exception form because
- adding the `check` argument to `call()` is awkward, because it looks a lot like the `env` and `cwd` arguments, that passthrough to `subprocess.run`, but there's a subtle difference - a different default. 
- adding the return type CompletedProcess to `call()` is also awkward, because it's only useful when `check=False`. 
- it muddles `call()`'s purpose, it's no longer 'do this' it's now 'do this/try this and tell me what happened'

So I thought the exception was neater overall.

[the remark this answers is not in the document]
Comment (henryiii) [reply cibuildwheel/macos.py:242]: Our entire wrapper is a bit bothersome - the important thing it does is add logging. It also happens to modify the interface just a tiny bit (different name, check=True). This is heavily because it was a shortcut for `check_call`, but we've standardized on `run(check=True)`, which is the recommended method to use for 3.5+ code. So what about renaming the wrapper to `utils.run`, just passing `**kwargs`, and adding `check=True` for most of the uses? I'd rather make it look like something people are used to (subprocess.run), rather than save a few chars.

Now if it was dramatically better, ala [plumbum](https://plumbum.readthedocs.io/en/latest/), then it would be worth learning something new. But not for just a tiny small change.

Could be updated later, of course, doesn't have to be in this PR.

[the remark this answers is not in the document]
Comment (henryiii) [reply cibuildwheel/macos.py:242]: Points 2 and 3 are just inherited from `subprocess.run`. The interface might not be ideal, but it's in the stdlib.

[the remark this answers is not in the document]
Comment (mayeut) [reply cibuildwheel/macos.py:242]: Would using:
```
    call(["python", "-m", "ensurepip"], env=env, cwd="/tmp")
    call(
         ["python", "-m", "pip", "install", "--force-reinstall", "pip", *dependency_constraint_flags],
         env=env,
         cwd="/tmp",
     )
```
solve the issue of detecting if pip is installed in the first place ?

[the remark this answers is not in the document]
Comment (henryiii) [reply cibuildwheel/macos.py:242]: Would that ensure the defaultpip script was there? I think so?

[the remark this answers is not in the document]
Comment (mayeut) [reply cibuildwheel/macos.py:242]: Given that an actual update would end-up in the place where we want to be and starts by removing the existing installation of pip, I would guess so. Let's try this. @joerick, I'll push a commit, please do revert it if you don't like what you're seeing :-)

[the remark this answers is not in the document]
Comment (joerick) [reply cibuildwheel/macos.py]: I agree- if that works, I'm interested. Let's see.

**Review by henryiii — APPROVED** [review]
I'm happy if we can get rid of this internal pip copy! :)

**Comment by joerick** [comment]
Cool, that worked! I just want to check out test suite timings though, it occurs to me that this might be reinstalling pip on every test now, which could be adding some time that we could do without...

**Comment by henryiii** [comment]
Weren't we doing that before? The pinned pip was never the bundled pip.

**Comment by joerick** [comment]
> Weren't we doing that before? The pinned pip was never the bundled pip.

True, but when running our integration tests, each test reuses the same environment, so once the pinned pip is installed, it doesn't get reinstalled on each test.

|                   | master | a116c2e (uninstall) | 7cde8fb (--force-reinstall) |
|---|---|---|---|
| GHA Linux         | 19m | 24m | 22m | 
| GHA Mac           | 28m | 20m | 22m |
| GHA Windows       | 35m | 36m | 40m |
| Azure Mac         | 22m | 18m | 21m |
| Azure Windows 3.6 | 35m | 37m | 47m |
| Azure Windows 3.8 | 35m | 47m | 61m |

There's quite a lot of variance in these figures, so I'm not totally sure, but it does look like an increase... let me try adding back the check for installation_dir/bin/pip before doing the reinstall, maybe we'll see a decrease.

**Comment by joerick** [comment]
updated table with the last commit-

|                   | master | a116c2e (uninstall) | 7cde8fb (--force-reinstall) | 39302c7 (--force-reinstall with existence check) |
|---|---|---|---|---|
| GHA Linux         | 19m | 24m | 22m | 23m |
| GHA Mac           | 28m | 20m | 22m | 22m |
| GHA Windows       | 35m | 36m | 40m | 36m |
| Azure Mac         | 22m | 18m | 21m | 14m |
| Azure Windows 3.6 | 35m | 37m | 47m | 34m |
| Azure Windows 3.8 | 35m | 47m | 61m | 36m |

**Comment by joerick** [comment]
Looks like it's a bit faster with a check around the `--force-reinstall` step. @mayeut does this look okay to you still?

**Comment by mayeut** [comment]
That looks good promising this way.
I'd use the "if not installed" to set a flag variable to `--force-reinstall` with the default being `--upgrade`.
It would get rid of 1 call to `pip install` and avoid code duplication for all args but this one.
What do you think ?

**Comment by joerick** [comment]
Really nice idea 👍

File: cibuildwheel/macos.py [hunk 3]
```diff
@@ -232,8 +231,26 @@ def setup_python(
...
         sys.exit(1)
 
-    # install pip & wheel
-    call(["python", get_pip_script, *dependency_constraint_flags], env=env, cwd="/tmp")
+    # ensure pip is installed
+    call(["python", "-m", "ensurepip"], env=env, cwd="/tmp")
+
+    # upgrade to the version matching our constraints
+    # if necessary, reinstall it to ensure that it's installed as 'pip'
+    requires_reinstall = not (installation_bin_path / "pip").exists()
```
Comment (mayeut) [inline cibuildwheel/macos.py:239]: maybe move `python -m ensurepip` below `requires_reinstall` to get rid of this call as well ?

```python
requires_reinstall = not (installation_bin_path / "pip").exists()
if requires_reinstall:
    # ensure pip is installed
    call(["python", "-m", "ensurepip"], env=env, cwd="/tmp")
```

Comment (joerick) [reply cibuildwheel/macos.py:239]: hmm. yeah, I guess it would be a little faster? I thought 'ensurepip' was a quick noop when it's installed, but it still takes around a second on my machine. I'll try your suggestion.

**Comment by joerick** [comment]
GHA having issues - https://www.githubstatus.com/

**Comment by mayeut** [comment]
updated table with the last commit-

|                   | master | a116c2e (uninstall) | 7cde8fb (--force-reinstall) | 39302c7 (--force-reinstall with existence check) | fd94e77 (conditional ensurepip) |
|---|---|---|---|---|---|
| GHA Linux         | 19m | 24m | 22m | 23m | 24m |
| GHA Mac           | 28m | 20m | 22m | 22m | 21m |
| GHA Windows       | 35m | 36m | 40m | 36m | 32m |
| Azure Mac         | 22m | 18m | 21m | 14m | 14m |
| Azure Windows 3.6 | 35m | 37m | 47m | 34m | 41m |
| Azure Windows 3.8 | 35m | 47m | 61m | 36m | 34m |

**Comment by joerick** [comment]
updated table with the last commits (no code changes)-

|                   | master | a116c2e (uninstall) | 7cde8fb (--force-reinstall) | 39302c7 (--force-reinstall with existence check) | fd94e77 (conditional ensurepip) | 3a61a5a | 9b48a41 | 
|---|---|---|---|---|---|---|
| GHA Linux         | 19m | 24m | 22m | 23m | 24m | 20m | 20m |
| GHA Mac           | 28m | 20m | 22m | 22m | 21m | 17m | 26m |
| GHA Windows   | 35m | 36m | 40m | 36m | 32m | 30m | 29m |
| Azure Mac         | 22m | 18m | 21m | 14m | 14m | 14m | 16m |
| Azure Windows 3.6 | 35m | 37m | 47m | 34m | 41m | 37m | 45m |
| Azure Windows 3.8 | 35m | 47m | 61m | 36m | 34m | 29m | 46m |

In summary, there might be too much variance in the Azure figures to conclude anything, but the GHA WIndows ones seem fairly stable. Conversely the Mac figures are more stable on Azure! 

Either way, this seems good to go 👍

**Comment by joerick** [comment]
thanks for spotting that license note henry!
